### ISSUE TITLE
Chan config / options

### ISSUE THREAD

**Issue by vmihailenco** [go-redis/redis#1700]
For https://github.com/go-redis/redis/issues/1673

Add new method `Chan` that accepts config to change buffer size and ping frequency, e.g.

```go
type ChanConfig {
    BufferSize         int
    PingFrequency time.Duration
}

func (c *PubSub) Chan(cfg ChanConfig) <-chan *Message {
    ...
}
```

In v9 `Chan` should completely replace `ChannelSize` and `Channel`.

---

Alternatively we could use options instead of config:

```go
func (c *PubSub) Chan(opt ...ChanOption) <-chan *Message {
    ...
}

pubsub.Chan(redis.WIthChanSize(100), redis.WithChanPingFrequency(time.Second))
```

Such API looks better / cooler, but for new users it is harder to find all available config options...

Any thoughts?

**Comment by rbg** [comment]
i like this solution over the solution in #1674 

I also do not think that it is that hard to find config options.. esp. if they are documented :)

**Comment by monkey92t** [comment]
we can increase the frequency of ping after disconnection to recover as soon as possiblew:
how do you feel?

```go
for {
		timer.Reset(pingTimeout)
		select {
		case <-timer.C:
			err := c.Ping(ctx).Err()
			if err != nil {
				pingTimeout = 1 * time.Second
			} else {
				pingTimeout = configTimeout
			}
		}
	}
```

**Comment by PeerXu** [comment]
> ```go
> for {
> 		timer.Reset(pingTimeout)
> 		select {
> 		case <-timer.C:
> 			err := c.Ping(ctx).Err()
> 			if err != nil {
> -				pingTimeout = 1 * time.Second
> +				pingTimeout = retryTimeout
> 			} else {
> 				pingTimeout = configTimeout
> 			}
> 		}
> 	}
> ```

On ping failed, pingTimeout should be optional.

> ```go
> func (c *PubSub) Chan(opt ...ChanOption) <-chan *Message {
>    ...
> }
>
> pubsub.Chan(
> 	redis.WIthChanSize(100), 
> 	redis.WithChanPingFrequency(time.Second /* default ping freq */, time.Second /* retried ping freq */))
> ```

About WithChanPingFrequency definition

```go
func WithChanPingFrequency(timeout time.Duration, retryTimeout ...time.Duration)
```

If retryTimeout not set, should be assigned by timeout.

**Comment by monkey92t** [comment]
Are there any advantages to setting retryTimeout?  I think our goal is to restore the connection as soon as possible after disconnection.

**Comment by PeerXu** [comment]
When we does not need to reconnect to server as soon as possible, 
we can set a lower ping cycle frequency to avoid reconnect to server frequently.

**Comment by monkey92t** [comment]
I don’t understand your idea very well, if you don’t need it, you can close pubsub.

**Comment by rbg** [comment]
@PeerXu's use case is where subscribed messages are on a redis server and the client is on the end of an IOT (cell/LoRa) network.  Since it costs real dollars to send/recv a ping the ask is for a means to externalize those parameters such that the ping payload does not happen quite so often. I suspect a more traditional environment this is a meh.. i don't care..

**Comment by monkey92t** [comment]
Fixed by #1764
